overlord/booted.go

```diff
@@ -57,22 +60,56 @@ func SyncBoot() error {
 	kernelSnap, _ := bootloader.GetBootVar("snappy_kernel")
 	osSnap, _ := bootloader.GetBootVar("snappy_os")
 
-	installed, err := (&Overlord{}).Installed()
+	st := ovld.State()
+	st.Lock()
+	installed, err := snapstate.All(st)
 	if err != nil {
 		return fmt.Errorf("cannot run SyncBoot: %s", err)
 	}
 
-	overlord := &Overlord{}
-	for _, snap := range []string{kernelSnap, osSnap} {
-		name, revno := nameAndRevnoFromSnap(snap)
-		found := FindSnapsByNameAndRevision(name, revno, installed)
-		if len(found) != 1 {
-			return fmt.Errorf("cannot SyncBoot, expected 1 snap %q (revision %s) found %d", snap, revno, len(found))
-		}
-		if err := overlord.SetActive(found[0], true, nil); err != nil {
-			return fmt.Errorf("cannot SyncBoot, cannot make %s active: %s", found[0].Name(), err)
+	var tsAll []*state.TaskSet
+	for _, snapNameAndRevno := range []string{kernelSnap, osSnap} {
+		name, rev := nameAndRevnoFromSnap(snapNameAndRevno)
```

Comment:
This function seems quite optimistic in its implementation. Anything bogus on the values will blow it up harshly.